ledger-service/http-json/src/main/scala/com/digitalasset/http/json/JsonProtocol.scala

```diff
@@ -238,33 +239,56 @@ object JsonProtocol extends DefaultJsonProtocol with ExtraFormats {
     *  1. template IDs are required
     *  2. query key may be absent
     *  3. special error if you appear to have queried outside 'query'
+    *
+    *  This provides an (almost) consistent behavior when reading the 'templateIds' and
+    *  'query' fields. Further extra fields may be added by concrete implementations.
     */
-  implicit val GetActiveContractsRequestFormat: RootJsonReader[domain.GetActiveContractsRequest] = {
-    case class GACR(
+  private def requestJsonReader[Request](validExtraFields: Set[String])(
+      toRequest: (
+          OneAnd[Set, TemplateId.OptionalPkg],
+          Map[String, JsValue],
+          Map[String, JsValue],
+      ) => Request
+  ): RootJsonReader[Request] = {
+    final case class BaseRequest(
         templateIds: Set[domain.TemplateId.OptionalPkg],
         query: Option[Map[String, JsValue]],
     )
-    val validKeys = Set("templateIds", "query")
-    implicit val primitive: JsonReader[GACR] = jsonFormat2(GACR.apply)
+    val validKeys = Set("templateIds", "query") ++ validExtraFields
+    implicit val primitive: JsonReader[BaseRequest] = jsonFormat2(BaseRequest.apply)
     jsv => {
-      val GACR(tids, q) = jsv.convertTo[GACR]
+      val BaseRequest(tids, query) = jsv.convertTo[BaseRequest]
       val extras = jsv.asJsObject.fields.keySet diff validKeys
       if (extras.nonEmpty)
         deserializationError(
-          s"unsupported query fields ${extras}; likely should be within 'query' subobject"
+          s"unsupported query fields $extras; likely should be within 'query' subobject"
         )
-      tids.headOption.cata(
-        h => domain.GetActiveContractsRequest(OneAnd(h, tids - h), q getOrElse Map.empty),
+      val extraFields = jsv.asJsObject.fields.view.filterKeys(validExtraFields).toMap
+      val nonEmptyTids = tids.headOption.cata(
+        h => OneAnd(h, tids - h),
         deserializationError("search requires at least one item in 'templateIds'"),
       )
+      toRequest(nonEmptyTids, query.getOrElse(Map.empty), extraFields)
     }
   }
 
+  implicit val GetActiveContractsRequestFormat: RootJsonReader[domain.GetActiveContractsRequest] =
+    requestJsonReader(validExtraFields = Set.empty)((tids, query, _) =>
+      domain.GetActiveContractsRequest(tids, query)
+    )
+
+  implicit val SearchForeverQueryFormat: RootJsonReader[domain.SearchForeverQuery] = {
+    val OffsetKey = "query_offset"
```

Comment:
Initially this was just `offset` but clashing with the generic `offset` message was difficult to manage. If there are suggestions on how to cleanly manage it I'm happy to change this back to just `offset`.

Reply:
The difference between what?

Reply:
I think what I was trying to say is "difficulty" not "difference". I blame lack of sleep.

Reply:
If a query contains the field `offset` the HTTP-JSON API thinks it's a late or duplicate offset message and it discards it with a message.